transport/x/grpc/headers.go

```diff
@@ -22,82 +22,102 @@ package grpc
 }
 
-// populateTransportRequest will populate the Request with all reserved and application
-// headers from the MD.
-func populateTransportRequest(md metadata.MD, request *transport.Request) error {
-	caller, callerErr := getFromMetadata(md, callerHeader)
-	request.Caller = caller
-	encoding, encodingErr := getFromMetadata(md, encodingHeader)
-	request.Encoding = transport.Encoding(encoding)
-	service, serviceErr := getFromMetadata(md, serviceHeader)
-	request.Service = service
-	headers, headersErr := getApplicationHeaders(md)
-	request.Headers = headers
-	return multierr.Combine(callerErr, encodingErr, serviceErr, headersErr)
+// metadataToTransportRequest will populate the Request with all reserved and application
+// headers into a new Request, only not setting the Body field.
+func metadataToTransportRequest(md metadata.MD) (*transport.Request, error) {
+	request := &transport.Request{
+		Headers: transport.NewHeadersWithCapacity(md.Len()),
+	}
+	for header := range md {
+		header = transport.CanonicalizeHeaderKey(header)
+		value, err := getFromMetadata(md, header)
+		if err != nil {
+			return nil, err
+		}
+		switch header {
+		case grpcheader.CallerHeader:
+			request.Caller = value
+		case grpcheader.ServiceHeader:
+			request.Service = value
+		case grpcheader.ShardKeyHeader:
+			request.ShardKey = value
+		case grpcheader.RoutingKeyHeader:
+			request.RoutingKey = value
+		case grpcheader.RoutingDelegateHeader:
+			request.RoutingDelegate = value
+		case grpcheader.EncodingHeader:
+			request.Encoding = transport.Encoding(value)
+		default:
+			request.Headers = request.Headers.With(header, value)
+		}
+	}
+	return request, nil
 }
 
-// add headers into md as application headers
-// return error if md already has a key defined that is defined in headers
+// addApplicationHeaders adds the headers to md.
 func addApplicationHeaders(md metadata.MD, headers transport.Headers) error {
-	for key, value := range headers.Items() {
-		if err := addToMetadata(md, applicationHeaderPrefix+key, value); err != nil {
+	for header, value := range headers.Items() {
+		header = transport.CanonicalizeHeaderKey(header)
+		if grpcheader.IsReserved(header) {
+			return fmt.Errorf("cannot use reserved header in application headers: %s", header)
+		}
+		if err := addToMetadata(md, header, value); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
-// get application headers from md
-// return error if any application error has more than one value
+// getApplicationHeaders returns the headers from md without any reserved headers.
 func getApplicationHeaders(md metadata.MD) (transport.Headers, error) {
 	headers := transport.NewHeadersWithCapacity(md.Len())
-	for mdKey := range md {
-		key := strings.TrimPrefix(mdKey, applicationHeaderPrefix)
-		// not an application header
-		if key == mdKey {
+	for header := range md {
+		header = transport.CanonicalizeHeaderKey(header)
+		if grpcheader.IsReserved(header) {
```

Comment:
grpc-go does this validation itself, as it should (there's a lot of them http://www.grpc.io/docs/guides/wire.html)